sdks/java/io/cassandra/src/main/java/org/apache/beam/sdk/io/cassandra/ReadFn.java

```diff
@@ -0,0 +1,172 @@
+import com.datastax.driver.core.Session;
+import com.datastax.driver.core.Token;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.beam.sdk.io.cassandra.CassandraIO.Read;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Joiner;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+class ReadFn<T> extends DoFn<Read<T>, T> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ReadFn.class);
+
+  private transient Cluster cluster;
+
+  private transient Session session;
+
+  private transient Read<T> lastRead;
+
+  @ProcessElement
+  public void processElement(@Element Read<T> read, OutputReceiver<T> receiver) {
+    Session session = getSession(read);
+    Mapper<T> mapper = read.mapperFactoryFn().apply(session);
+    String partitionKey =
+        cluster.getMetadata().getKeyspace(read.keyspace().get()).getTable(read.table().get())
+            .getPartitionKey().stream()
+            .map(ColumnMetadata::getName)
+            .collect(Collectors.joining(","));
+
+    String query = generateRangeQuery(read, partitionKey, read.ringRanges() != null);
+    PreparedStatement preparedStatement = session.prepare(query);
+    Set<RingRange> ringRanges =
+        read.ringRanges() == null ? Collections.<RingRange>emptySet() : read.ringRanges().get();
+
+    for (RingRange rr : ringRanges) {
+      Token startToken = cluster.getMetadata().newToken(rr.getStart().toString());
+      Token endToken = cluster.getMetadata().newToken(rr.getEnd().toString());
+      ResultSet rs =
+          session.execute(preparedStatement.bind().setToken(0, startToken).setToken(1, endToken));
+      Iterator<T> iter = mapper.map(rs);
+      while (iter.hasNext()) {
+        T n = iter.next();
+        receiver.output(n);
+      }
+    }
+
+    if (read.ringRanges() == null) {
+      ResultSet rs = session.execute(preparedStatement.bind());
+      Iterator<T> iter = mapper.map(rs);
+      while (iter.hasNext()) {
+        receiver.output(iter.next());
+      }
+    }
+  }
+
+  @Teardown
+  public void teardown() {
+    if (session != null) {
+      this.session.close();
+    }
+    if (cluster != null) {
+      this.cluster.close();
+    }
+  }
+
+  private Session getSession(Read<T> read) {
+    if (cluster == null || !reuseCluster(this.lastRead, read)) {
+      this.cluster =
+          CassandraIO.getCluster(
+              read.hosts(),
+              read.port(),
+              read.username(),
+              read.password(),
+              read.localDc(),
+              read.consistencyLevel());
+    }
+    if (session == null || !reuseSession(lastRead, read)) {
+      this.session = this.cluster.connect(read.keyspace().get());
+    }
+    this.lastRead = read;
+    return this.session;
+  }
+
+  private static <T> boolean reuseCluster(Read<T> readA, Read<T> readB) {
+    return readA != null
```

Comment:
This is still a bit messy, don't you think we can get something similar by using something like connection pooling and delegate this complexity to the driver (where it should reside)?
https://docs.datastax.com/en/developer/java-driver/2.1/manual/pooling/

Notice that I am not familiar with this but after a quicklook seems to be worth the look.

Reply:
I didn't address this yet, wanted your thoughts on it.  I agree it can be improved. 

The driver will handle Connection *pooling*, (handling N number of connections per cassandra host and sharing them in a thread safe way), but we still need to initiate a `Session` per Cassandra instance, and maintain a reference to the `Session` to keep those connections in scope (under the hood).  

So we could use something as simple as a `Map<?, Cluster>` and `Map<?, Session>`, but remember we have to do both, since each `Read<>` might be hitting an entirely different cassandra cluster.   If we wanted something fancier, we could try a guava LRU or something on the off chance the user is initiating hundreds(thousands/millions?) of different Session objects.